engine/consensus/sealing/core.go

```diff
@@ -448,30 +448,44 @@ func (c *Core) processApproval(approval *flow.ResultApproval) error {
 }
 
 func (c *Core) checkEmergencySealing(observer consensus.SealingObservation, lastSealedHeight, lastFinalizedHeight uint64) error {
+	// if emergency sealing is not activated, then exit
 	if !c.config.EmergencySealingActive {
 		return nil
 	}
 
-	emergencySealingHeight := lastSealedHeight + approvals.DefaultEmergencySealingThreshold
+	// emergency sealing is checked at collector level, however we don't have to check collector level if
+	// there are not so many blocks are not sealed.
+	// Since emergency sealing won't trigger unless there are at least DefaultEmergencySealingThresholdForExecution blocks finalized
+	// after it, we can use this fact to skip checking at collector level
 
-	// we are interested in all collectors that match condition:
-	// lastSealedBlock + sealing.DefaultEmergencySealingThreshold < lastFinalizedHeight
-	// in other words we should check for emergency sealing only if threshold was reached
-	if emergencySealingHeight >= lastFinalizedHeight {
+	if lastSealedHeight+approvals.DefaultEmergencySealingThresholdForExecution >= lastFinalizedHeight {
 		return nil
 	}
 
-	delta := lastFinalizedHeight - emergencySealingHeight
-	// if block is emergency sealable depends on it's incorporated block height
-	// collectors tree stores collector by executed block height
-	// we need to select multiple levels to find eligible collectors for emergency sealing
-	for _, collector := range c.collectorTree.GetCollectorsByInterval(lastSealedHeight, lastSealedHeight+delta) {
-		err := collector.CheckEmergencySealing(observer, lastFinalizedHeight)
-		if err != nil {
-			return err
-		}
+	// prevent from underflow in case lastFinalizedHeight is too small
+	if lastFinalizedHeight < approvals.DefaultEmergencySealingThresholdForExecution {
+		return nil
 	}
-	return nil
+
+	// We iterate through the collectorTree to check if any collector is able to trigger emergency sealing.
+	// We can do some optimization instead of bindly iterate through all collectors.
+	// The optimization is to start from the collector for the first unsealed block, if it can trigger emergency sealing, we
+	// check for the next unsealed and finalized block, and stop as soon as we met with a height that cannot trigger emergency sealing.
+	// The reason this optimization could work is because the following facts:
+	// 1) emergency sealing is only for finalized blocks,
+	// 2) for finalized blocks, there is only one collector at each height
```

Comment:
I am not sure why this should hold. Imagine we have an execution fork. Then there would be two different results (in the finalized blocks) both awaiting approvals from verifiers.

Reply:
sorry, yeah, you are right. I realized later as well.

My optimization could work with execution fork on finalized blocks, just forgot to update the comments here. Will update

Reply:
I think you would need to change the logic to 
```
if none of the results for this block height can be sealed, then we don't need to check descending results whether they can be emergency sealed. 
```